File: ios/Sources/EventLog.swift
import Foundation
import SwiftUI

/// A single timestamped entry in the global event log.
struct LogEntry: Identifiable {
    let id: Int
    let ts: String
    let msg: String
}

/// Global event log shared across every tab — mirrors the React Native
/// reference app's `addLog` / `useLogs`. SDK calls and route changes append
/// here; the Profile tab's debug sheet renders the full list.
@MainActor
final class EventLog: ObservableObject {
    static let shared = EventLog()

    @Published private(set) var entries: [LogEntry] = []
    private var nextId = 0

    private static let formatter: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "HH:mm:ss"
        return f
    }()

    private init() {}

    /// Append a message. Safe to call from any thread/context.
    nonisolated func add(_ msg: String) {
        if Thread.isMainThread {
            MainActor.assumeIsolated { self._add(msg) }
        } else {
            Task { @MainActor in self._add(msg) }
        }
    }

    private func _add(_ msg: String) {
        nextId += 1
        let entry = LogEntry(id: nextId, ts: Self.formatter.string(from: Date()), msg: msg)
        entries.insert(entry, at: 0)
        if entries.count > 100 {
            entries = Array(entries.prefix(100))
        }
    }
}

File: flutter/rum/lib/event_log.dart
import 'package:flutter/foundation.dart';

/// A single entry in the global event log.
class LogEntry {
  LogEntry(this.id, this.ts, this.msg);

  final int id;
  final String ts;
  final String msg;
}

/// Global, app-wide event log shared across every screen.
///
/// Mirrors the reference app's module-level `addLog` + `useLogs` pattern: SDK
/// calls, route changes, and network outcomes append here, and the Profile
/// tab's debug sheet renders the list reactively via the [ValueNotifier].
class EventLog {
  EventLog._();

  static final ValueNotifier<List<LogEntry>> entries =
      ValueNotifier<List<LogEntry>>(<LogEntry>[]);

  static int _nextId = 0;

  /// Append a message to the log (most-recent-first, capped at 100 entries).
  static void add(String msg) {
    final DateTime now = DateTime.now();
    final String ts =
        '${_two(now.hour)}:${_two(now.minute)}:${_two(now.second)}';
    final LogEntry entry = LogEntry(++_nextId, ts, msg);
    final List<LogEntry> next = <LogEntry>[
      entry,
      ...entries.value.take(99),
    ];
    entries.value = next;
  }

  static String _two(int v) => v.toString().padLeft(2, '0');
}

/// Convenience top-level alias mirroring the reference's `addLog(...)`.
void addLog(String msg) => EventLog.add(msg);
